src/components/Tooltip.js

```diff
@@ -0,0 +1,183 @@
+import _ from 'underscore';
+import React, {Component} from 'react';
+import PropTypes from 'prop-types';
+import {Animated, Text, View} from 'react-native';
+import Hoverable from './Hoverable';
+import withWindowDimensions, {windowDimensionsPropTypes} from './withWindowDimensions';
+import getTooltipStyles from '../styles/getTooltipStyles';
+
+const propTypes = {
+    // The text to display in the tooltip.
+    text: PropTypes.string.isRequired,
+
+    // Children to wrap with Tooltip.
+    children: PropTypes.node.isRequired,
+
+    // Props inherited from withWindowDimensions
+    ...windowDimensionsPropTypes,
+};
+
+class Tooltip extends Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            // The distance between the left side of the wrapper view and the left side of the window
+            xOffset: 0,
+
+            // The distance between the top of the wrapper view and the top of the window
+            yOffset: 0,
+
+            // The width and height of the wrapper view
+            wrapperWidth: 0,
+            wrapperHeight: 0,
+
+            // The width and height of the tooltip itself
+            tooltipWidth: 0,
+            tooltipHeight: 0,
+        };
+
+        // The wrapper view containing the wrapped content along with the Tooltip itself.
+        this.wrapperView = null;
+
+        // The tooltip (popover) itself.
+        this.tooltip = null;
+
+        this.animation = new Animated.Value(0);
+
+        this.getWrapperPosition = this.getWrapperPosition.bind(this);
+        this.measureWrapperAndGetPosition = this.measureWrapperAndGetPosition.bind(this);
+        this.measureTooltip = this.measureTooltip.bind(this);
+        this.showTooltip = this.showTooltip.bind(this);
+        this.hideTooltip = this.hideTooltip.bind(this);
+    }
+
+    shouldComponentUpdate(nextProps, nextState) {
+        /*
+           We only want to re-render if props or state have changed.
+           This prevents an infinite rendering loop with the `onLayout` callback calling setState.
+           Also, we need to do a deep comparison of props using _.isEqual rather than relying on the shallow comparison
+           done by `PureComponent` because a shallow comparison would not catch changes in `props.windowDimensions`.
```

Comment:
NAB. 

This makes sense, but it does have me wondering about why the infinite rendering loop is happening and whether there is no better way to prevent it.

Some questions I have...
- In what scenario do we need to re-measure the items? 
- What exactly causes a "layout" to occur?
- Is `onLayout` typically called on every `render()`?

Separately, we could maybe get a bit more specific about the conditions where we want to allow a render (e.g. "only re-render when `(this.state.xOffset !== nextState.xOffset) || etc`").

Not sure we need to block on this per se, but I am curious to learn more about what is happening with `onLayout`.

Reply:
> In what scenario do we need to re-measure the items

It's difficult to know when we need to re-measure the wrapper view, because it could be anything. It could be completely static, or it could be constantly changing – we don't really know what the children are doing. At minimum, we need to measure the wrapper view initially and when the window dimensions change.

The tooltip would only need to be re-measured if `this.props.text` changes.

> What exactly causes a "layout" to occur

I'm really not sure ... I couldn't find the answer to this question in the docs.

> Is onLayout typically called on every render()

I thought so, but maybe not ... according to the docs, `onLayout` is "Invoked on mount and layout changes".

> Separately, we could maybe get a bit more specific about the conditions where we want to allow a render

I originally had this fully-expanded like you're saying, but I think we would want a re-render to occur if any state or prop changes:

[ ] - `this.props.text` - Yes because this would change the tooltip width and height
[ ] - `this.props.children` - Yes because this would change the wrapper width, height, maybe position too
[ ] - `this.props.windowDimensions` - Yes because this would make `this.state.xOffset` and `this.state.yOffset` incorrect.
[ ] - `this.state.xOffset` and `this.state.yOffset` - Yes because otherwise the tooltip will display offset from the component.
[ ] - `this.state.XXXwidth` and `this.state.XXXheight` - Yes because the updated width and height (resulting from the change in `this.props.text` or `this.props.children` is not available until after the next `render` -> `onLayout` -> `render`

Reply:
Thanks for exploring that. Cool with this for now, but still curious if there really is any "infinite rendering" problem here (which would seem to indicate that `onLayout` is getting called every render) and if so whether there's anyway around it. 

Another future improvement here might be to refactor `withWindowDimensions` so it returns less complex props (something like `width`, `height`, and `isSmallScreen` instead of an object). Then we can use `PureComponent` or `React.memo()` here to solve this problem.

Reply:
Cool, NAB for now so I'm going to resolve